Fix incorrect unit conversion factors in units.py

raulloiscuns opened the pull request:
<!-- 
Insert issue number that this PR fixes (if any) just after 'Fixes #'.
If this PR does not fix an existing issue, consider opening one or remove 'Fixes #' from the PR description.
-->
Fixes #5051

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 
The conversion factors for A/fs, A/ms and A/us have been revised and corrected in `units.py`.

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [X] Issue raised/referenced?
 - [X] Tests updated/added?
 - [x] Documentation updated/added?
 - [X] `package/CHANGELOG` file updated?
 - [X] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5053.org.readthedocs.build/en/5053/

<!-- readthedocs-preview mdanalysis end -->

raulloiscuns committed:
Change in speed conversion factors

raulloiscuns committed:
Fix unit conversion factors in units.py and update corresponding test cases

raulloiscuns committed:
Merge branch 'develop' of https://github.com/raulloiscuns/mdanalysis into develop

raulloiscuns committed:
Fix speed unit conversion factors in CHANGELOG

raulloiscuns committed:
Update velocity formatting in LAMMPS output and adjust reference velocities for test consistency

orbeckst commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -365,7 +365,7 @@ def _write_velocities(self, atoms):
         for index, vel in zip(indices, velocities):
             self.f.write(
-                "{i:d} {x:f} {y:f} {z:f}\n".format(
+                "{i:d} {x:.8f} {y:.8f} {z:.8f}\n".format(
```
This may be useful but does not belong into this PR. Please revert and make a separate PR – then people knowledgable about LAMMPS can discuss.

orbeckst replied:
reverted

orbeckst replied:
It seems that the writer test does not pass unless we force higher precision. Our LAMMPS test files have 10 decimals so I am tentatively setting all output to `.10f`.

orbeckst commented on package/MDAnalysis/units.py:
```diff
@@ -301,23 +301,23 @@ def __getitem__(self, key):
+    "Angstrom/femtosecond": 1.0 * 1e-3,
+    "angstrom/femtosecond": 1.0 * 1e-3,
+    "angstrom/fs": 1.0 * 1e-3,
+    "A/fs": 1.0 * 1e-3,
```
Could you please also add Å/ns (1.0 * 1e3) – for whatever reason we never added that one.

orbeckst replied:
And add an entry to CHANGELOG under Enhancements 
```
* add conversion factor for speed unit A/ns (PR #5053)
```

orbeckst replied:
added (and reorganized for better readability)

orbeckst commented on package/CHANGELOG:
please also add your GitHub username to the list at ??/??/?? so that we see who contributed to the 2.10.0 release.

orbeckst replied:
added

orbeckst commented on package/CHANGELOG:
```diff
@@ -33,6 +33,8 @@ Fixes
  * Fixes the benchmark `SimpleRmsBench` in `benchmarks/analysis/rms.py`
    by changing the way weights for RMSD are calculated, instead of
    directly passing them. (Issue #3520, PR #5006)
+ * Fixes bug in `units.py`: conversion factors for some speed units has
+   been fixed (Issue #5051)  
```
Let's be explicit:
```suggestion
 * Fixes incorrect conversion factors for speed units A/fs, A/us, A/ms 
   in MDAnalysis.units (Issue #5051, PR #5053)  
```

orbeckst commented on testsuite/MDAnalysisTests/utils/test_units.py:
```diff
@@ -134,9 +134,9 @@ def test_force(self, quantity, unit1, unit2, ref):
-            (1, "A/fs", "A/ps", 1e-3),
+            (1, "A/ms", "A/ps", 1e-9),
+            (1, "A/us", "A/ps", 1e-6),
+            (1, "A/fs", "A/ps", 1e3),
```
add a test for A/nm, please

orbeckst replied:
added tests (including other missing speed units)

orbeckst reviewed (changes requested):
Thank you VERY much for finding this very old bug. I assume you found it while working with LAMMPS files? Great work fixing and also updating the tests.

Please address the following points (see inline comments for details)
- update CHANGELOG (include your GitHub handle and rewrite/add entries a bit)
- _please_ add Å/nm to the conversions (I will merge your PR if you don't do it but it would be nice as it fits directly in with your fix)
- please remove the formatting change to the LAMMPS writer, we'd like to keep PRs focused on one thing

Apologies that it took a while to review your PR.

orbeckst commented:
@raulloiscuns it would be great to add your fix to MDAnalysis soon. Are you able to address the review comments?

orbeckst committed:
Merge branch 'develop' into develop

orbeckst committed:
Merge branch 'develop' into develop

orbeckst committed:
Update package/CHANGELOG

orbeckst committed:
Merge branch 'develop' into develop

orbeckst committed:
updated CHANGELOG

orbeckst committed:
reverted formatting of v in LAMMPS DATAWriter + flake8 fixes

orbeckst committed:
add A/ns to speed units

- updated table of units
- added tests

orbeckst committed:
blackified MDAnalysis.units

orbeckst reviewed (commented):
I am finishing the PR so that we can include it in the 2.10.0 release.

orbeckst committed:
blackened test_units.py

IAlibay commented on package/CHANGELOG:
```diff
@@ -37,19 +36,22 @@ Fixes
  * Fixes the benchmark `SimpleRmsBench` in `benchmarks/analysis/rms.py`
    by changing the way weights for RMSD are calculated, instead of
    directly passing them. (Issue #3520, PR #5006)
+ * Fixes incorrect conversion factors for speed units A/fs, A/us, A/ms
```
Please put things done in this PR at the top of the entry list.

orbeckst replied:
ok, I'll also move the entry from PR #5113 up

IAlibay reviewed (commented):
I'm not reviewing this PR, but please do order the changelog entries properly to save me time on release.

orbeckst commented:
I am using pre-commit with black 24 and it passes and then our linters/black fails; I probably need to sync the settings in some way, until then there'll be more micro commits of the type "blackened file xxx" (which I'll squash merge).

orbeckst committed:
move CHANGELOG entries for fixes to newest first order

IAlibay commented on package/MDAnalysis/coordinates/LAMMPS.py:
```diff
@@ -832,9 +830,8 @@ def _read_next_timestep(self):
-                f"No coordinates following convention "
-                "{self.lammps_coordinate_convention} found in "
-                "timestep"
+                "No coordinates following convention "
```
This tells me this error was not tested, otherwise it should have thrown an error for misusing f-strings - could you raise an issue to follow-up on this?

orbeckst replied:
flake8 found it.

orbeckst replied:
Do we need a test for it?? I opened https://github.com/MDAnalysis/mdanalysis/issues/5127

IAlibay replied:
> need a test

I'd say yes, but not in scope of this PR - it's a good first issue thing.

IAlibay commented on package/MDAnalysis/units.py:
```diff
@@ -301,23 +302,28 @@ def __getitem__(self, key):
+    "A/fs": 1.0 * 1e-3,
+    "Angstrom/ns": 1.0 * 1e3,
+    "A/ns": 1.0 * 1e3,
+    "\u212b/ns": 1.0 * 1e3,
```
no entry for "\u212b/fs"?

IAlibay replied:
or microsecond, or millisecond, etc...

orbeckst replied:
I added a few others and the ones you mentione. I can add more but it makes little sense to be combinatorically complete. This is just old code and I am not particularly proud of it ;-).

IAlibay replied:
Thanks!

IAlibay commented on package/CHANGELOG:
Not this PR, just a general comment. We should probably come up with some kind of rule about autoformatting the changelog, personally not sure it makes sense to have black apply to it since it's just free-form text.

orbeckst replied:
Sorry, didn't see that it removed all the trailing whitespace.

This is my fault for not configuring pre-commit hook properly (or disabling it). It won't let me commit until it's happy and I may just not have the right `.pre-commit-config.yaml`.

orbeckst replied:
I _think_ it wasn't black that touched CHANGELOG because when I only committed CHANGELOG, black did not engage
```
(mda-develop) mac:MDAnalysisTests oliver$ git commit --amend
black................................................(no files to check)Skipped
flake8...............................................(no files to check)Skipped
trim trailing whitespace.................................................Passed
fix end of files.........................................................Passed
check yaml...........................................(no files to check)Skipped
check for added large files..............................................Passed
check for merge conflicts................................................Passed
```
Instead it's likely the *trim trailing whitespace* plugin for pre-commit.

It would be great to have an expert share a pre-commit yaml file for MDA development.

IAlibay commented on testsuite/MDAnalysisTests/utils/test_units.py:
```diff
@@ -134,9 +134,14 @@ def test_force(self, quantity, unit1, unit2, ref):
-            (1, "A/fs", "A/ps", 1e-3),
+            (1, "A/ms", "A/ps", 1e-9),
+            (1, "A/us", "A/ps", 1e-6),
+            (1, "A/fs", "A/ps", 1e3),
```
Could you add at least one test for some of those unicode entries? just to make sure it actually works?

orbeckst replied:
added (not exhaustive)

IAlibay reviewed (changes requested):
I might as well actually review rather than just lurk.

IAlibay commented:
> I am using pre-commit with black 24 and it passes and then our linters/black fails; I probably need to sync the settings in some way, until then there'll be more micro commits of the type "blackened file xxx" (which I'll squash merge).

cc @RMeli

orbeckst committed:
more speed-unit enhancements

- add tests for unit conversion with unicode characters
- add additional representations Å/us, Å/µs, Å/fs, Å/ms

orbeckst commented on package/MDAnalysis/coordinates/LAMMPS.py:
Sorry for the noise. When I reverted a change, pre-commit/flake/black were all unhappy about the file so I did some line breaking and a single real fix.

orbeckst committed:
set output precision in LAMMPS DATAWriter to .10f for x and v

- match precision in our test files
- required for tests to pass when round-tripping velocities
- updated CHANGELOG

IAlibay reviewed (approved):
lgtm, thanks!

orbeckst committed:
blackened utils/test_units.py

orbeckst commented:
Thanks @IAlibay , just updated CHANGELOG with output precision changes and ran black locally again. I think I just don't have the right settings for pre-commit black and our black and need to sort that out.

If the latest changes pass our CI, I'll also approve and squash-merge. Thanks a lot for your extra 👀 — much appreciated!

orbeckst commented:
(This got automerged on approval... I had _thought_ it would wait for _all_ of CI... oops, sorry. All linters and Linux had passed at least. If anything gets messed up on develop I'll fix it.)

orbeckst commented:
@raulloiscuns your PR got merged and your fix will be in 2.10.0. 🎉  Thank you very much for identifying the problem and proposing a solution!

RMeli commented:
Sorry @orbeckst and @IAlibay, I just noticed the ping. But looks like the linters passed, do you know what changed in order to fix things?

orbeckst commented:
I am using the following `.pre-commit.yml`
<details>

```yaml
# Pre-commit configuration for MDAnalysis
repos:
  # Black - Python code formatter
  - repo: https://github.com/psf/black
    # workaround https://github.com/psf/black/issues/2493#issuecomment-1081987650
    rev: 'refs/tags/24.10.0:refs/tags/24.10.0'
    hooks:
      - id: black
        language_version: python3
        # Use project's Black configuration from pyproject.toml
        files: \.(py)$

  # Flake8 - Python linter (more lenient for this project)
  - repo: https://github.com/pycqa/flake8
    rev: 7.3.0
    hooks:
      - id: flake8
        args: [
          --max-line-length=88,
          "--extend-ignore=E203,W503,F401,F841,E722,E711,E712,E713,E714,F811"
        ]
        files: \.(py)$

  # Additional useful hooks
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v6.0.0
    hooks:
      - id: trailing-whitespace
      - id: end-of-file-fixer
      - id: check-yaml
      - id: check-added-large-files
      - id: check-merge-conflict
```
</details>

and I would have _hoped_ that it fixes anything that our GH linters would find offensive. Instead, I had to repeatedly manually run `black` on files that the black linter did not like.

I probably didn't configure my pre-commit exactly the way that MDAnalysis wants it. A sample yaml file that "just works" would be useful.
